components/cors-mgt/org.wso2.carbon.identity.cors.mgt.core/src/main/java/org/wso2/carbon/identity/cors/mgt/core/internal/impl/CORSManagementServiceImpl.java

```diff
@@ -0,0 +1,313 @@
+    }
+
+    /**
+     * Create a ResourceTypeAdd for CORS Origin type.
+     *
+     * @return ResourceTypeAdd A resource type with CORS_ORIGIN_RESOURCE_TYPE set as the name.
+     */
+    private ResourceTypeAdd createCORSOriginResourceTypeToAdd() {
+
+        ResourceTypeAdd resourceTypeAdd = new ResourceTypeAdd();
+        resourceTypeAdd.setName(CORS_ORIGIN_RESOURCE_TYPE_NAME);
+        resourceTypeAdd.setDescription(CORS_ORIGIN_RESOURCE_TYPE_DESCRIPTION);
+        return resourceTypeAdd;
+    }
+
+    /**
+     * Retrieve the ConfigurationManager instance from the CORSServiceHolder.
+     *
+     * @return ConfigurationManager The ConfigurationManager instance.
+     */
+    private ConfigurationManager getConfigurationManager() {
+
+        return CORSManagementServiceHolder.getInstance().getConfigurationManager();
+    }
+
+    /**
+     * Returns true if the tenant already has a particular CORS Origin.
+     *
+     * @param origin The Origin to be checked against the existing Origins.
+     * @return {@code true} if the tenant already have the particular CORS Origin, {@code false} otherwise.
+     * @throws ConfigurationManagementException
+     */
+    private boolean tenantHasCORSOrigin(CORSOrigin origin) throws ConfigurationManagementException {
+
+        Resource resource = getConfigurationManager().getResource(CORS_ORIGIN_RESOURCE_TYPE_NAME,
+                CORS_ORIGIN_RESOURCE_NAME);
+        if (resource != null) {
+            List<CORSOrigin> currentCORSOrigins = new ResourceToCORSOrigin().apply(resource);
+
+            return currentCORSOrigins.contains(origin);
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Validate the CORSOrigin list.
+     *
+     * @param corsOrigins List of CORSOrigin instances.
+     * @throws CORSManagementServiceClientException
+     */
+    private void validateOrigins(List<CORSOrigin> corsOrigins) throws CORSManagementServiceClientException {
+        if (corsOrigins == null) {
+            if (log.isDebugEnabled()) {
+                log.debug(ERROR_CODE_EMPTY_LIST.getMessage());
+            }
+            throw handleClientException(ERROR_CODE_EMPTY_LIST);
+        }
+
+        for (CORSOrigin corsOrigin : corsOrigins) {
+            if (isInvalidOrigin(corsOrigin)) {
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format(ERROR_CODE_INVALID_ORIGIN.getMessage(), corsOrigin.getUrl()));
+                }
+                throw handleClientException(ERROR_CODE_INVALID_ORIGIN, corsOrigin.getUrl());
+            }
+        }
+    }
+
+    /**
+     * Check if the format of the Origin is valid.
+     *
+     * @param origin Origin to be checked for validity.
+     * @return {@code true} if the origin is valid, {@code false} otherwise.
+     */
+    private boolean isInvalidOrigin(CORSOrigin origin) {
+
+        try {
+            new URL(origin.getUrl()).toURI();
+        } catch (MalformedURLException | URISyntaxException e) {
```

Comment:
let's add a debug log to log the exception

Reply:
This is logged in the higher-level function where `isInvalidOrigin` is called.